File: src/test/java/com/uwetrottmann/trakt5/services/SeasonsTest.java
package com.uwetrottmann.trakt5.services;

import com.uwetrottmann.trakt5.BaseTestCase;
import com.uwetrottmann.trakt5.TestData;
import com.uwetrottmann.trakt5.entities.Episode;
import com.uwetrottmann.trakt5.entities.Ratings;
import com.uwetrottmann.trakt5.entities.Season;
import com.uwetrottmann.trakt5.entities.Stats;
import com.uwetrottmann.trakt5.enums.Extended;
import org.junit.Test;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class SeasonsTest extends BaseTestCase {

    @Test
    public void test_summary() throws IOException {
        List<Season> seasons = executeCall(getTrakt().seasons().summary(TestData.SHOW_SLUG,
        Extended.FULLEPISODES));
        assertThat(seasons).isNotNull();
        assertThat(seasons).hasSize(5);
        for (Season season : seasons) {
            assertThat(season).isNotNull();
            // must have at least trakt and tvdb id
            assertThat(season.ids.trakt).isPositive();
            if (season.ids.tvdb != null) {
                assertThat(season.ids.tvdb).isPositive();
            }
            assertThat(season.title).isNotNull();
            assertThat(season.network).isNotNull();
            // seasons start at 0 for specials
            assertThat(season.number).isGreaterThanOrEqualTo(0);
            assertThat(season.episode_count).isPositive();
            assertThat(season.aired_episodes).isGreaterThanOrEqualTo(0);
            assertThat(season.rating).isBetween(0.0, 10.0);
            assertThat(season.votes).isGreaterThanOrEqualTo(0);
            // episode details
            if (season.number == TestData.EPISODE_SEASON) {
                assertThat(season.episodes).isNotNull();
                assertThat(season.episodes).isNotEmpty();
                Episode firstEp = null;
                for (Episode episode : season.episodes) {
                    if (episode.number == TestData.EPISODE_NUMBER) {
                        firstEp = episode;
                        break;
                    }
                }
                assertThat(firstEp).isNotNull();
                assertThat(firstEp.title).isEqualTo(TestData.EPISODE_TITLE);
                assertThat(firstEp.season).isEqualTo(TestData.EPISODE_SEASON);
                assertThat(firstEp.number).isEqualTo(TestData.EPISODE_NUMBER);
                assertThat(firstEp.ids.imdb).isEqualTo(TestData.EPISODE_IMDB_ID);
                assertThat(firstEp.ids.tmdb).isEqualTo(TestData.EPISODE_TMDB_ID);
                assertThat(firstEp.ids.tvdb).isEqualTo(TestData.EPISODE_TVDB_ID);
                assertThat(firstEp.overview).isNotEmpty();
            }
        }
    }

    @Test
    public void test_season() throws IOException {
        List<Episode> season = executeCall(getTrakt().seasons().season(TestData.SHOW_SLUG, TestData.EPISODE_SEASON,
                null));
        assertThat(season).isNotNull();
        assertThat(season).isNotEmpty();
        for (Episode episode : season) {
            assertThat(episode.season).isEqualTo(TestData.EPISODE_SEASON);
        }
    }

    @Test
    public void test_comments() throws IOException {
        executeCall(getTrakt().seasons().comments(TestData.SHOW_SLUG, TestData.EPISODE_SEASON));
    }

    @Test
    public void test_ratings() throws IOException {
        Ratings ratings = executeCall(getTrakt().seasons().ratings(TestData.SHOW_SLUG, TestData.EPISODE_SEASON));
        assertRatings(ratings);
    }

    @Test
    public void test_stats() throws IOException {
        Stats stats = executeCall(getTrakt().seasons().stats(TestData.SHOW_SLUG, TestData.EPISODE_SEASON));
        assertShowStats(stats);
    }

}

File: src/test/kotlin/com/redissi/trakt/services/SeasonsTest.kt
package com.redissi.trakt.services

import com.redissi.trakt.BaseTestCase
import com.redissi.trakt.TestData
import com.redissi.trakt.TestRatings
import com.redissi.trakt.TestShowStats
import com.redissi.trakt.entities.Episode
import com.redissi.trakt.enums.Extended
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.*
import org.junit.Test
import java.io.IOException

class SeasonsTest : BaseTestCase(), TestRatings, TestShowStats {
    @Test
    @Throws(IOException::class)
    fun `get all seasons for a show`() = runBlocking {
        val seasons = trakt.seasons().summary(TestData.SHOW_SLUG, Extended.FULLEPISODES)
        seasons.shouldNotBeNull().shouldHaveSize(6)
        for (season in seasons) {
            season.shouldNotBeNull()
            // must have at least trakt and tvdb id
            season.ids.shouldNotBeNull()
            season.ids!!.trakt.shouldNotBeNull().shouldBePositive()
            if (season.ids!!.tvdb != null) {
                season.ids!!.tvdb.shouldNotBeNull().shouldBePositive()
            }
            season.title.shouldNotBeNull()
            season.network.shouldNotBeNull()
            // seasons start at 0 for specials
            season.number.shouldNotBeNull().shouldBeGreaterOrEqualTo(0)
            season.episodeCount.shouldNotBeNull().shouldBePositive()
            season.airedEpisodes.shouldNotBeNull().shouldBeGreaterOrEqualTo(0)
            season.rating.shouldNotBeNull().shouldBeInRange(0.0..10.0)
            season.votes.shouldNotBeNull().shouldBeGreaterOrEqualTo(0)
            // episode details
            if (season.number == TestData.EPISODE_SEASON) {
                season.episodes.shouldNotBeNull().shouldNotBeEmpty()
                var firstEp: Episode? = null
                for (episode in season.episodes!!) {
                    if (episode.number == TestData.EPISODE_NUMBER) {
                        firstEp = episode
                        break
                    }
                }
                firstEp.shouldNotBeNull()
                firstEp.title.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_TITLE)
                firstEp.season.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_SEASON)
                firstEp.number.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_NUMBER)
                firstEp.ids.shouldNotBeNull()
                firstEp.ids!!.imdb.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_IMDB_ID)
                firstEp.ids!!.tmdb.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_TMDB_ID)
                firstEp.ids!!.tvdb.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_TVDB_ID)
                firstEp.overview.shouldNotBeNullOrEmpty()
            }
        }
    }

    @Test
    @Throws(IOException::class)
    fun `get a single season for a show`() = runBlocking {
        val season = trakt.seasons().season(TestData.SHOW_SLUG, TestData.EPISODE_SEASON, null)
        season.shouldNotBeNull().shouldNotBeEmpty()
        for (episode in season) {
            episode.season.shouldNotBeNull().shouldBeEqualTo(TestData.EPISODE_SEASON)
        }
    }

    @Test
    @Throws(IOException::class)
    fun `get season comments`() = runBlocking<Unit> {
        val comments = trakt.seasons().comments(TestData.SHOW_SLUG, TestData.EPISODE_SEASON)
        comments.shouldNotBeNull()
        comments.size.shouldBeLessOrEqualTo(DEFAULT_PAGE_SIZE)
    }

    @Test
    @Throws(IOException::class)
    fun `get season ratings`() = runBlocking {
        val ratings = trakt.seasons().ratings(TestData.SHOW_SLUG, TestData.EPISODE_SEASON)
        ratings.shouldNotBeNull()
        assertRatings(ratings)
    }

    @Test
    @Throws(IOException::class)
    fun `get season stats`() = runBlocking {
        val stats = trakt.seasons().stats(TestData.SHOW_SLUG, TestData.EPISODE_SEASON)
        stats.shouldNotBeNull()
        assertShowStats(stats)
    }
}